[FLINK-2373] Add configuration parameter to createRemoteEnvironment method

akunft opened the pull request:
Adds an overloaded ExecutionEnvironment.createRemoteEnvironment method to provide a custom configuration for the job client. This resolves problems with the default configuration values, e.g. large jar files causing errors when they exceeded the default payload size of Akka.

tillrohrmann commented on flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:
```diff
@@ -1105,6 +1105,27 @@ public static ExecutionEnvironment createRemoteEnvironment(String host, int port
+	 * to a cluster for execution. Note that all file paths used in the program must be accessible from the
+	 * cluster. The custom configuration file is used to configure the default parallelism and Akka specific
+	 * configuration parameters for the JobClient only. Cluster configuration has to be done in the remotely
+	 * running Flink instance.
```
Very good :-)

tillrohrmann commented:
Very good contribution. Thanks @akunft. The only thing would be to add a test which verifies that the configuration is actually used for starting the `JobClientActor's` `ActorSystem`, for example. We could set the framesize to something so small that the job execution must fail. Or we change the default parallelism as it is done in `ExecutionEnvironmnetITCase.testLocalEnvironmentWithConfig`.

After adding the test, +1 for merging :-)

uce commented:
Thanks for the PR. I agree with Till.

@tillrohrmann, regarding #1016: what would be the way to use the RemoteEnvironment with an HA cluster? This will already work with HA, but the hostname and port would be ignored, right? Should Andreas already add a variant with config and jar files only? Or is that something we should address with your PR?

tillrohrmann commented:
@uce, I think this should be addressed either in #1016 or in a separate PR once #1016 is merged.

akunft commented:
After thinking about it, I would change the behavior so that the config is only for defining akka specific settings. The default parallelism of the operators can be set via ```setParallelism```, just like in the local environment. Otherwise, we had to use ```parallelism.default``` to set the operators parallelism, which is confusing imo. This also goes in line with the configuration description on the website.

What do you think? 

In any case, I will create unit tests and update the PR once we decided on a version.

akunft commented:
Added tests + behavior described above.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/RemoteEnvironmentITCase.java:
```diff
@@ -0,0 +1,166 @@
+        DataSet<String> result = env.createInput(new TestNonRichInputFormat());
+        result.output(new LocalCollectionOutputFormat<String>(new ArrayList<String>()));
+        try {
+            env.execute();
```
Shouldn't we fail after this statement?

akunft replied:
Yes, I just ensure its cause of the wrong akka setting.

tillrohrmann replied:
I meant shouldn't we insert an `Assert.fail()` after this statement. If we successfully execute `execute` then something wrong happened, right?

akunft replied:
It would still fail, as no exception is thrown. But I can generate a nicer error message. Will fix that.

tillrohrmann replied:
Sorry, you're completely right. My bad.

On Wed, Sep 2, 2015 at 6:54 PM, Andreas Kunft <notifications@github.com>
wrote:

> In
> flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/RemoteEnvironmentITCase.java
> <https://github.com/apache/flink/pull/1066#discussion_r38556479>:
>
> > +    @Test(expected=IllegalArgumentException.class)
> > +    public void testInvalidAkkaConfiguration() throws Throwable {
> > +        Configuration config = new Configuration();
> > +        config.setString(ConfigConstants.AKKA_STARTUP_TIMEOUT, INVALID_STARTUP_TIMEOUT);
> > +
> > +        final ExecutionEnvironment env = ExecutionEnvironment.createRemoteEnvironment(
> > +                cluster.hostname(),
> > +                cluster.getLeaderRPCPort(),
> > +                config
> > +        );
> > +        env.getConfig().disableSysoutLogging();
> > +
> > +        DataSet<String> result = env.createInput(new TestNonRichInputFormat());
> > +        result.output(new LocalCollectionOutputFormat<String>(new ArrayList<String>()));
> > +        try {
> > +            env.execute();
>
> It would still fail, as no exception is thrown. But I can generate a nicer
> error message. Will fix that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/1066/files#r38556479>.
>

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/RemoteEnvironmentITCase.java:
```diff
@@ -0,0 +1,166 @@
+            return 1;
+        }
+    }
+}
```
This file is indented with whitespaces

akunft replied:
done

tillrohrmann commented:
From your last comment I conclude that you only wanted to allow akka specific features to be set with the configuration. However, I couldn't find how you check this. Furthermore, if I remove the `env.setParallelism` call in the `testUserSpecificParallelism` test case, then I can still control the default DOP with `config.setInteger(ConfigConstants.DEFAULT_PARALLELISM, 42)`, for example.

akunft commented:
Yes, I didn't enforce that the config parameter is not used, but the setParallelism overrides it.
I let it like this, as it is more confusing to the user if all this is explained in the java docs and maybe someone wants to set it in the config.

fhueske commented:
LGTM, @tillrohrmann do you want to have another look as well?

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/RemoteEnvironmentITCase.java:
```diff
@@ -0,0 +1,164 @@
+					}
+				});
+		List<Integer> resultCollection = new ArrayList<Integer>();
+		result.output(new LocalCollectionOutputFormat<Integer>(resultCollection));
```
Here you can also use `result.collect()` to obtain a list of integers.

tillrohrmann commented:
LGTM modulo one minor comment.

akunft commented:
* rebased
* replaced collection output with ```collect()``` in RemoteEnvironmentITCase and ExecutionEnvironmentITCase
* added ```createRemoteEnvironment``` method with config parameter in Scala ExecutionEnvironment

fhueske commented:
Thanks for the update!
Will merge this PR tomorrow.
